projects/batfish-common-protocol/src/test/java/org/batfish/storage/FileStorageImplTest.java

```diff
@@ -0,0 +1,145 @@
+
+    //Test equality of persistent and transient instances
+    Assert.assertThat(persistedCopy, equalTo(analysis));
+  }
+
+  @Test
+  public void testInvalidSaveAnalysis() {
+    String containerName = "testcontainer";
+    String analysisNmae = "testanalysis";
+    Map<String, String> question =
+        Collections.singletonMap("testquestionname", "testquestionvalue");
+    Analysis analysis = new Analysis(analysisNmae, question);
+    _storage.saveAnalysis(containerName, analysis);
+
+    //Test that save analysis throws for duplicate analysis
+    _thrown.expect(BatfishException.class);
+    _thrown.expectMessage(
+        String.format(
+            "Analysis '%s' already exists for container '%s'", analysisNmae, containerName));
+    _storage.saveAnalysis(containerName, analysis);
+  }
+
+  @Test
+  public void testUpdateAnalysis() {
+    Analysis analysis =
+        new Analysis(
+            "testanalysisupdate", Collections.singletonMap("testquestion", "testquestionvalue"));
+    _storage.saveAnalysis("testcontainerupdate", analysis);
+    Analysis savedAnalysis = _storage.getAnalysis("testcontainerupdate", "testanalysisupdate");
+    savedAnalysis.addQuestion("modifiedquestion", "modifiedquestionvalue");
+    Analysis updatedAnalysis = _storage.updateAnalysis("testcontainerupdate", savedAnalysis);
+
+    //test that attributes were updated
+    assertThat(updatedAnalysis, equalTo(savedAnalysis));
+  }
+
+  @Test
+  public void testInvalidUpdateAnalysis() {
+    Analysis newAnalysis =
+        new Analysis(
+            "testanalysisupdate", Collections.singletonMap("testquestion", "testquestionvalue"));
+    _thrown.expect(BatfishException.class);
+    _thrown.expectMessage(
+        String.format(
+            "Analysis '%s' doesn't exist for container '%s'",
+            "testanalysisupdate", "testcontainer"));
+
+    //test that update throws for non-existent analysis
+    _storage.updateAnalysis("testcontainer", newAnalysis);
+  }
+
+  @Test
+  public void testDeleteEmptyAnalysis() {
+    Analysis analysis = new Analysis("testanalysis", new HashMap<>());
+    _storage.saveAnalysis("testcontainer", analysis);
+    _storage.deleteAnalysis("testcontainer", "testanalysis", false);
+    _thrown.expect(BatfishException.class);
+    _thrown.expectMessage(
+        String.format(
+            "Analysis '%s' doesn't exist for container '%s'", "testanalysis", "testcontainer"));
+
+    //should throw for the deleted analysis
+    _storage.getAnalysis("testcontainer", "testanalysis");
+  }
+
+  @Test
+  public void testDeleteNonEmptyAnalysis() {
+    Analysis analysis =
+        new Analysis(
+            "testanalysis", Collections.singletonMap("testquestion", "testquestioncontent"));
+    _storage.saveAnalysis("testcontainer", analysis);
+    _thrown.expect(BatfishException.class);
+    _thrown.expectMessage(
+        String.format("%s' is not empty, deletion must be forced", "testanalysis"));
```

Comment:
There should be a leading '? Suggest to use _thrown.expectMessage(equalsTo()) for comparing message equality, default only checks the actual message contains the parameter string as subString.

Reply:
I am not sure about this comment, as Stack Overflow https://stackoverflow.com/questions/2469911/how-do-i-assert-my-exception-message-with-junit-test-annotation and existing code in BatfishTest and other classes do the same